### PR TITLE
Dev deepfm multicol oneemb

### PR DESCRIPTION
**Pull request by Liuxinman**

### CODE REVIEW

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+    eval_time = time.time() - eval_start_time
+    
+    auc = 0  # will be updated by rank 0 only
+    rank = flow.env.get_rank()
+    if rank == 0:
+        auc_start_time = time.time()
+        auc = flow.roc_auc_score(labels, preds).numpy()[0]
+        auc_time = time.time() - auc_start_time
+        log_loss_start_time = time.time()
+        logloss = log_loss(labels, preds, eps=1e-7)
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:583]: replace `log_loss` with `flow._C.binary_cross_entropy_loss`

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+        )
+
+    return auc, logloss
+
+
+if __name__ == "__main__":
+    os.system(sys.executable + " -m oneflow --doctor")
+    flow.boxing.nccl.enable_all_to_all(True)
+    args = get_args()
+    train(args)
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:604]: add a blank line

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+                        + f"Latency {(latency * 1000):0.3f} ms, Throughput {throughput:0.1f}, {strtime}"
+                    )
+
+            if step % batches_per_epoch == 0:
+                epoch += 1
+                auc, logloss = eval(args, eval_graph, tag='val', cur_step=step, epoch=epoch, cached_eval_batches=cached_eval_batches)
+                if args.save_model_after_each_eval:
+                    save_model(f"step_{step}_val_auc_{auc:0.5f}")
+
+                if rank == 0:
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:488]: All devices should perform early stop.

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+        .to_local()
+    )
+
+    flow.comm.barrier()
+    eval_time = time.time() - eval_start_time
+    
+    auc = 0  # will be updated by rank 0 only
+    logloss = 0 # will be updated by rank 0 only
+    rank = flow.env.get_rank()
+    if rank == 0:
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:579]: all devices should calculate auc and log_loss, but only rank 0 print loss and so on.

File: RecommenderSystems/deepfm/train_deepfm_criteo_x4.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      --loss_print_interval 100 \
+      --dnn "1000,1000,1000,1000,1000" \
+      --net_dropout 0.2 \
+      --learning_rate 0.001 \
+      --embedding_vec_size 16 \
+      --num_train_samples 36672493 \
+      --num_val_samples 4584062 \
+      --num_test_samples 4584062 \
+      --model_save_dir $MODEL_SAVE_DIR \
+      > run.log & tail -f run.log
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/train_deepfm_criteo_x4.sh:35]: use `tee` or nothing

File: RecommenderSystems/deepfm/deepfm_train_eval.py (hunk 3 above)
Comment (Liuxinman) [reply RecommenderSystems/deepfm/deepfm_train_eval.py]: fixed

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+                if save_best:
+                    print(f"Save best model: monitor(max): {best_metric:.6f}")
+                    save_model("best_checkpoint")
+                if stop_training:
+                    break
+                
+                deepfm_module.train()
+                last_time = time.time()
+
+    print(f"Loading model from {args.model_save_dir}/best_checkpoint")
```
[the remark this answers is not in the document]
Comment (Liuxinman) [reply RecommenderSystems/deepfm/deepfm_train_eval.py]: Add a load_model function and check if path is valid before loading

File: RecommenderSystems/deepfm/train_deepfm_criteo_x4.sh (hunk 5 above)
Comment (Liuxinman) [reply RecommenderSystems/deepfm/train_deepfm_criteo_x4.sh]: deleted

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+                    if rank == 0:
+                        print(f"Save best model: monitor(max): {best_metric:.6f}")
+                    save_model("best_checkpoint")
+                if stop_training:
+                    break
+                
+                deepfm_module.train()
+                last_time = time.time()
+    
+    load_model(f"{args.model_save_dir}/best_checkpoint")
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:519]: check if model save dir is available

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 8]
```diff
@@ -0,0 +1,615 @@
...
+
+        # FM
+        lr_out = flow.sum(lr_embedded_x, dim=1)
+        dot_sum = interaction(embedded_x)
+        fm_pred = lr_out + dot_sum
+
+        # DNN
+        dnn_pred = self.dnn_layer(embedded_x.flatten(start_dim=1))
+
+        return self.final_activation(fm_pred + dnn_pred)
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:329]: may use tensor.sigmoid()?

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+
+    def forward(self, x: flow.Tensor) -> flow.Tensor:
+        return self.linear_layers(x)
+
+
+def interaction(embedded_x:flow.Tensor) -> flow.Tensor:
+    sum_of_square = flow.sum(embedded_x, dim=1) ** 2
+    square_of_sum = flow.sum(embedded_x ** 2, dim=1)
+    bi_interaction = (sum_of_square - square_of_sum) * 0.5
+    return flow.sum(bi_interaction, dim=-1).view(-1, 1)
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:279]: return flow.sum(bi_interaction, dim=-1, keepdim=True)

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+        
+        self.final_activation = nn.Sigmoid()
+
+    def forward(self, inputs) -> flow.Tensor:
+        multi_embedded_x = self.embedding_layer(inputs)
+        embedded_x = multi_embedded_x[:, :, 0:self.embedding_vec_size]
+        lr_embedded_x = multi_embedded_x[:, :, -1].unsqueeze(-1)
+
+        # FM
+        lr_out = flow.sum(lr_embedded_x, dim=1)
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:322]: lr_embedded_x = multi_embedded_x[:, :, -1]
lr_out = flow.sum(lr_embedded_x, dim=1, keepdim=True)

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 11]
```diff
@@ -0,0 +1,608 @@
...
+
+    def get_batches(self, reader, batch_size=None):
+        if batch_size is None:
+            batch_size = self.batch_size
+
+        tail = None
+
+        for rg in reader:
+            rgdict = rg._asdict()
+            rglist = np.array([rgdict[field] for field in self.fields])
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:140]: np.array will force to convert dtype to float64. 

```python
        for rg in reader:
            rgdict = rg._asdict()
            rglist = [rgdict[field] for field in self.fields]
            pos = 0
            if tail is not None:
                pos = batch_size - len(tail[0])
                tail = list(
                    [
                        np.concatenate((tail[i], rglist[i][0 : (batch_size - len(tail[i]))]))
                        for i in range(self.num_fields)
                    ]
                )
                if len(tail[0]) == batch_size:
                    label = tail[0]
                    features = tail[1 : self.num_fields]
                    tail = None
                    yield label, np.stack(features, axis=-1)
                else:
                    pos = 0
                    continue
            while (pos + batch_size) <= len(rglist[0]):
                label = rglist[0][pos : pos + batch_size]
                features = [rglist[j][pos : pos + batch_size] for j in range(1, self.num_fields)]
                pos += batch_size
                yield label, np.stack(features, axis=-1)
            if pos != len(rglist[0]):
                tail = [rglist[i][pos:] for i in range(self.num_fields)]

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 12]
```diff
@@ -0,0 +1,609 @@
...
+                last_time = time.time()
+    
+    load_model(f"{args.model_save_dir}/best_checkpoint")
+    if rank == 0:
+        print("================ Test Evaluation ================")
+    eval(args, eval_graph, tag='test', cur_step=step, epoch=epoch)
+
+
+def np_to_global(np, dtype=flow.float):
+    # TODO: t = flow.from_numpy(np)
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:525]: from_numpy should work

Comment (Liuxinman) [reply RecommenderSystems/deepfm/deepfm_train_eval.py:525]: 暂时还不行，从fuxi .h5 load出来的数据是nparray，数据类型都是float64，我明天看看重新做一下数据集

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 13]
```diff
@@ -0,0 +1,673 @@
...
+        default=1024,
+        help="size of cache memory budget on each device in megabytes when store_type is cached_host_mem or cached_ssd",
+    )
+
+    parser.add_argument(
+        "--amp", action="store_true", help="enable Automatic Mixed Precision(AMP) training or not"
+    )
+    parser.add_argument("--loss_scale_policy", type=str, default="static", help="static or dynamic")
+
+    parser.add_argument("--early_stop", action="store_true", help="enable early stop or not")
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:105]: disable_early_stop

File: RecommenderSystems/deepfm/train_deepfm_criteo_x4.sh [hunk 14]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+export CUDA_VISIBLE_DEVICES=2
+export ONEFLOW_DEBUG_MODE=INFO
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/train_deepfm_criteo_x4.sh:3]: remove these two exports.

**Review by ShawnXuan — COMMENTED** [review]
RecommenderSystems/deepfm/tools/criteo_parquet.py this file is not ready, suggest to remove it in this pr.

File: RecommenderSystems/deepfm/README.md [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+     ```
+
+     After generating parquet dataset, dataset information will also be printed. It contains the information about the number of samples and table size array, which is needed when training.
+
+      ```txt
+     train samples = 36672493                                                             
+     validation samples = 4584062
+     test samples = 4584062                                                               
+     table size array: 
+     15,68,56,36,173,93,43,41,114,5,16,6,44
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/README.md:115]: suggest to use the real values here, in case somebody copy them.

**Review by ShawnXuan — APPROVED** [review]

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 16]
```diff
@@ -0,0 +1,675 @@
...
+            persistent_path=persistent_path,
+            table_size_array=table_size_array,
+            store_type=one_embedding_store_type,
+            cache_memory_budget_mb=cache_memory_budget_mb,
+            size_factor=3,
+        )
+
+        self.dnn_layer = DNN(
+            in_features=embedding_vec_size * (num_dense_fields + num_sparse_fields),
+            hidden_units=dnn + [1],
```
Comment (liujuncheng) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:361]: 这里的 `1` 是不是应该作为o ut_features 传过去

Comment (Liuxinman) [reply RecommenderSystems/deepfm/deepfm_train_eval.py:361]: 可以

**Review by liujuncheng — APPROVED** [review]
